### PR TITLE
Jetty 9.4.x 2043 concurrentmodification annotation parsing

### PR DESCRIPTION
**Pull request by janbartel**

### CODE REVIEW

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java [hunk 1]
```diff
@@ -561,31 +564,26 @@ public AnnotationParser(int javaPlatform, int asmVersion)
      */
     public void addParsedClass (String classname, Resource location)
     {
-        List<String> list = new ArrayList<>(1);
-        if (location != null)
-            list.add(location.toString());
-
-        List<String> existing = _parsedClassNames.putIfAbsent(classname, list);
+        String existing = _parsedClassNames.putIfAbsent(classname, location.toString());
```
Comment (sbordet) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: I would save the call to `toString()` here, and use a `Map<String, Resource>`.

File: jetty-annotations/src/test/java/org/eclipse/jetty/annotations/TestAnnotationParser.java [hunk 2]
```diff
@@ -71,6 +74,36 @@ public void handle(ClassInfo info, String annotation)
...
         }
     }
+    
+    
+    public static class DuplicateClassScanHandler extends AnnotationParser.AbstractHandler
+    {
+        private Map<String, List<String>> _classMap = new ConcurrentHashMap();
+
+        /** 
+         * @see org.eclipse.jetty.annotations.AnnotationParser.AbstractHandler#handle(org.eclipse.jetty.annotations.AnnotationParser.ClassInfo, java.lang.String)
```
Comment (sbordet) [inline jetty-annotations/src/test/java/org/eclipse/jetty/annotations/TestAnnotationParser.java]: Remove this javadoc, it's boilerplate that is not necessary (IDEs will present the javadocs of the super method).

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java (hunk 1 above)
Comment (janbartel) [reply jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: But the point is to output the log warning, which wants the resource as a string. So I've solved it a different way by doing location.toString() once.

File: jetty-annotations/src/test/java/org/eclipse/jetty/annotations/TestAnnotationParser.java (hunk 2 above)
Comment (janbartel) [reply jetty-annotations/src/test/java/org/eclipse/jetty/annotations/TestAnnotationParser.java]: Well, my IDE inserted the comment :)  I've deleted it.

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java [hunk 3]
```diff
@@ -71,7 +71,10 @@
     private static final Logger LOG = Log.getLogger(AnnotationParser.class);
     protected static int ASM_OPCODE_VERSION = Opcodes.ASM6; //compatibility of api
     
-    protected Map<String, List<String>> _parsedClassNames = new ConcurrentHashMap<>();
+    /**
+     * Map of classnames scanned and the first location from which scan occurred
+     */
+    protected Map<String, String> _parsedClassNames = new ConcurrentHashMap<>();
```
Comment (sbordet) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: I think this should be `Map<String, Resource>`. See also below.

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java [hunk 4]
```diff
@@ -561,31 +564,27 @@ public AnnotationParser(int javaPlatform, int asmVersion)
      */
     public void addParsedClass (String classname, Resource location)
     {
-        List<String> list = new ArrayList<>(1);
-        if (location != null)
-            list.add(location.toString());
-
-        List<String> existing = _parsedClassNames.putIfAbsent(classname, list);
+        String locationStr = location.toString();
+        String existing = _parsedClassNames.putIfAbsent(classname, locationStr);
```
Comment (sbordet) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: What I meant in the previous review was:

```java
String existing = _parsedClassNames.putIfAbsent(classname, location);
```

File: jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java [hunk 5]
```diff
@@ -561,31 +564,27 @@ public AnnotationParser(int javaPlatform, int asmVersion)
...
+        
         if (existing != null)
-        {
-            existing.addAll(list);
-            LOG.warn("{} scanned from multiple locations: {}", classname, existing);
-        }
+            LOG.warn("{} scanned from multiple locations: {}, {}", classname, existing, locationStr);
```
Comment (sbordet) [inline jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: Revert back to log `location`, not `locationStr`.

Comment (janbartel) [reply jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: But that's 2 toStrings! One for the autobox to String on the putIfAbsent, and then the other on location in the LOG.warn.

Comment (janbartel) [reply jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java]: OK ok. You win. Change made.

**Pull request merged**